src/main/java/uk/ac/cam/cl/dtg/isaac/api/AssignmentFacade.java

```diff
@@ -939,105 +935,149 @@ public Response getAssignedGroupsByGameboards(@Context final HttpServletRequest
             }
 
-            if (null == assigneeGroup) {
-                return new SegueErrorResponse(Status.BAD_REQUEST, "The group id specified does not exist.")
-                        .toResponse();
+            // Assert that there is at least one assignment, and that multiple assignments are only set by staff
+            if (assignmentDTOsFromClient.size() == 0) {
+                return new SegueErrorResponse(Status.BAD_REQUEST, "You need to specify at least one assignment to set.").toResponse();
+            } else if (!userIsStaff && assignmentDTOsFromClient.size() > 1) {
+                return new SegueErrorResponse(Status.FORBIDDEN, "You need a staff account to set assignments to more than one group at once!").toResponse();
             }
 
-            if (!GroupManager.isOwnerOrAdditionalManager(assigneeGroup, currentlyLoggedInUser.getId())
-                    && !isUserAnAdmin(userManager, currentlyLoggedInUser)) {
-                return new SegueErrorResponse(Status.FORBIDDEN,
-                        "You can only set assignments to groups you own or manage.").toResponse();
-            }
+            // Since all of the assignment notes in a single query are currently the same, we only have to do note size
+            // checks once, and we only need to fetch the gameboard once
+            AssignmentDTO firstAssignment = assignmentDTOsFromClient.get(0);
+            boolean notesIsNullOrEmpty = firstAssignment.getNotes() == null || (firstAssignment.getNotes() != null && firstAssignment.getNotes().isEmpty());
+            boolean notesIsTooLong = firstAssignment.getNotes() != null && firstAssignment.getNotes().length() > MAX_NOTE_CHAR_LENGTH;
 
             if (userIsStaff) {
                 if (notesIsTooLong) {
-                    return new SegueErrorResponse(Status.BAD_REQUEST, "Your assignment notes exceed the maximum allowed length of " +
-                            MAX_NOTE_CHAR_LENGTH.toString() + " characters.").toResponse();
+                    return new SegueErrorResponse(Status.BAD_REQUEST, "Your assignment notes exceed the maximum allowed length of "
+                            + MAX_NOTE_CHAR_LENGTH.toString() + " characters.").toResponse();
                 }
             } else if (!notesIsNullOrEmpty) {
                 // user is not staff but it is a teacher, if we got here unscathed
                 return new SegueErrorResponse(Status.BAD_REQUEST, "You are not allowed to add assignment notes.").toResponse();
             }
 
-            GameboardDTO gameboard = this.gameManager.getGameboard(assignmentDTOFromClient.getGameboardId());
+            if (firstAssignment.getGameboardId() == null) {
+                return new SegueErrorResponse(Status.BAD_REQUEST, "A required field was missing. Must provide gameboard id.").toResponse();
+            }
+
+            GameboardDTO gameboard = this.gameManager.getGameboard(firstAssignment.getGameboardId());
```

Comment:
What if each AssignmentDTO has a different gameboard? Again just because the frontend won't let you doesn't mean it won't happen.

Reply:
This validation is important for the later loop not to fail, so this also needs to be inside the loop (although I'd recommend a cache of valid gameboards to avoid looking up the same one again and again).